src/main/java/com/google/sps/data/PersistentUserRepository.java

```diff
@@ -0,0 +1,179 @@
+    }
+
+    datastore.put(userEntity);
+  }
+
+  // function to fetch the user from the database
+  public Collection<User> fetchUserProfiles() {
+    Query query = new Query("User");
+    PreparedQuery results = datastore.prepare(query);
+
+    Collection<User> userProfiles = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+        long id = entity.getKey().getId();
+        String name = (String) entity.getProperty("name");
+        Collection<String> specialties = (Collection<String>) entity.getProperty("specialties");
+
+        User userObject = new User(name);
+        if(specialties != null) {
+            for (String specialty : specialties) {
+                userObject.addSpecialty(specialty);
+            }
+        }
+        if(specialties != null) {
+            for (String specialty : specialties) {
+                userObject.addSpecialty(specialty);
+            }
+        }
+        userObject.setID(id);
+        userProfiles.add(userObject);
+    }
+    return userProfiles;
+  }
+
+  // function that sets a query filter for the results in the datastore
+  public PreparedQuery getQueryFilterForUsers(String name) {
+    Filter userNameFilter = new FilterPredicate("name", FilterOperator.EQUAL, name);
+    Query query = new Query("User").setFilter(userNameFilter);
+    PreparedQuery results = datastore.prepare(query);
+    return results;
+  }
+
+    // function that fetches a single user, collection of users, filter on that user for name?
+  public Collection<User> fetchSingleUserProfile(User user) {
+    String inputUserName = user.getName();
+    PreparedQuery results = getQueryFilterForUsers(inputUserName);
+
+    Collection<User> userProfiles = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+        String name = (String) entity.getProperty("name");
+
+        User userObject = new User(name);
+        
+        userProfiles.add(userObject);
+    }
+    return userProfiles;
+  }
+
+  // function that adds user to the database if it does not exist already
+  public void addUser(User user) {
+    Collection<User> userProfiles = fetchSingleUserProfile(user);
+    /*
+     * if the userProfiles is larger than 1, that means the 
+     * if it doesnt exist then it gets added
+    */
+    if(userProfiles.size() == 0) {
+        addUserToDatabase(user);
+    }
+  }
+
+  public void removeUserProfile(User user) throws Exception{
+    String inputUserName = user.getName();
+    System.out.println(inputUserName);
+    PreparedQuery results = getQueryFilterForUsers(inputUserName);
+    int size = results.countEntities();
+    /*
+     *  if size is greater than 0 then there are entities available,
+     *  if size is 0 then users do not equal, so an exception is thrown
+    */
+    if(size > 0){
+        for (Entity entity : results.asIterable()) {
+            String currentName = (String) entity.getProperty("name");
+            if(inputUserName.contains(currentName)) {
```

Comment:
Using contains here may be error prone for similar reasons as your previous commits:

If the inputUserName is "joseph campbell" and there exists an entity whose name is "joe", then "joe" and "joseph campbell" would both get deleted.

Reply:
I see. Would .matches() method be better? The tests still pass if I use matches

Reply:
Would making it say this be better:

currentName.matches(inputUserName) 
?

Reply:
I used .equals()

Reply:
Done